### PR TITLE
dotCMS/core#23335 Error when edit a container

### PR DESCRIPTION
**Pull request by zulqarnainvd**
### Proposed Changes
* remove console error
* set values correctly in FormGroup
* change the Architecture of the dot-properties component

### Checklist
- [x] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)

### Additional Info
** any additional useful context or info **

### Screenshots
Original             |  Updated
:-------------------------:|:-------------------------:
** original screenshot **  |  ** updated screenshot **

### CODE REVIEW

**Review by jdotcms — APPROVED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.html [hunk 1]
```diff
@@ -19,27 +19,30 @@
             </ng-template>
         </p-tabPanel>
         <p-tabPanel
-            *ngFor="let containerContent of containerContents.controls; let i = index"
-            [header]="containerContent.value.name"
+            *ngFor="let containerContent of getcontainerStructures.controls; let i = index"
+            [header]="getContentTypeNameById(containerContent.value.structureId)"
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.html:23]: https://plainenglish.io/blog/angular-dont-call-a-function-inside-the-template-anymore

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        //
+    }
+}
+
+const messageServiceMock = new MockDotMessageService({
+    'containers.properties.add.variable.title': 'Title',
+    Add: 'Add'
+});
+
+describe('DotContentEditorComponent', () => {
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.spec.ts:126]: You are not testing the core functionality of this component here.

You need to tests all the add and remove content types.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts [hunk 3]
```diff
@@ -92,27 +58,14 @@ export class DotContentEditorComponent implements ControlValueAccessor, OnInit {
...
-        });
-    }
-
     /**
      * It removes the form control at the index of the form array, and then closes the modal
      * @param {number} [index=null] - number = null
      * @param close - This is the function that closes the modal.
      * @memberof DotContentEditorComponent
      */
     handleClose(index: number = null, close: () => void): void {
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts:67]:
```suggestion
    removeItem(index: number = null, close: () => void): void {
```

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts [hunk 4]
```diff
@@ -129,17 +82,30 @@ export class DotContentEditorComponent implements ControlValueAccessor, OnInit {
...
                         .setValue(editor.getValue() + `${variable}`);
                 }
             }
         });
     }
 
+    /**
+     * Get content type name
+     * @param {string} id
+     * @return {*}
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts:99]: The doc is wrong.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts [hunk 5]
```diff
@@ -162,24 +128,27 @@ export class DotContentEditorComponent implements ControlValueAccessor, OnInit {
         this.activeTabIndex = index;
     }
 
-    private mapMenuItems(contentTypes: DotContainerContent[]): MenuItem[] {
+    private mapMenuItems(contentTypes: DotCMSContentType[]): MenuItem[] {
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts:131]:
```suggestion
    private getMenuItems(contentTypes: DotCMSContentType[]): MenuItem[] {
```

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts [hunk 6]
```diff
@@ -162,24 +128,27 @@ export class DotContentEditorComponent implements ControlValueAccessor, OnInit {
...
                 }
             };
         });
     }
 
-    private checkIfAlreadyExists(contentType: DotContainerContent): boolean {
-        return this.containerContents.controls.some(
-            (control) => control.value.variable === contentType.variable
+    private checkIfAlreadyExists(contentType: DotCMSContentType): boolean {
+        return this.getcontainerStructures.controls.some(
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts:150]: Are you sure that `some` is the right method for this task?

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts [hunk 7]
```diff
@@ -158,5 +209,37 @@ describe('DotContainerPropertiesComponent', () => {
...
+
+        it('should setup code', () => {
+            const field = de.query(By.css('[data-testId="code"]'));
+            expect(field).toBeDefined();
+        });
+
+        it('should button enable when max content greater then zero', fakeAsync(() => {
+            fixture.componentInstance.form.get('maxContentlets').setValue(2);
+            const contentTypeButton = de.query(By.css('[data-testId=showContentTypeAndCode]'));
+            contentTypeButton.nativeElement.click();
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts:236]: https://angular.io/guide/testing-components-scenarios#clicking

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts [hunk 8]
```diff
@@ -158,5 +209,37 @@ describe('DotContainerPropertiesComponent', () => {
...
+            tick(3000);
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts:237]: This is a clear sign this component is bad implemented, it should not take 3 seconds to show the component with the code and the tabs.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.ts [hunk 9]
```diff
@@ -59,16 +54,46 @@ export class DotContainerPropertiesComponent implements OnInit {
...
             this.store.updateIsContentTypeButtonEnabled(values.maxContentlets > 0);
         });
     }
 
+    /**
+     * Map Container Strcutures into FormControl
+     * @param {DotContainerStructure[]} containerStructures
+     * @memberof DotContainerPropertiesComponent
+     */
+    mapContainerStructureIntoFormControl(containerStructures: DotContainerStructure[]) {
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.ts:74]:
```suggestion
    addContainerFormControl(containerStructures: DotContainerStructure[]) {
```

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-loop-editor/dot-loop-editor.component.spec.ts [hunk 10]
```diff
@@ -88,16 +88,16 @@ fdescribe('DotLoopEditorComponent', () => {
...
+    it('should show pre_loop', () => {
         const label = de.query(By.css('[data-testId="label"]')).nativeElement;
         expect(label.innerText).toBe('Pre-loop');
     });
 
-    fit('should show pre_loop', fakeAsync(() => {
+    it('should show pre_loop', fakeAsync(() => {
         component.label = 'post_loop';
         fixture.detectChanges();
         tick();
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-loop-editor/dot-loop-editor.component.spec.ts:103]: Why this need tick and fake async?

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-loop-editor/dot-loop-editor.component.spec.ts (hunk 10 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-loop-editor/dot-loop-editor.component.spec.ts:103]: I removed the `fakeAsync` but when I run the test previously it's failing but now it's working fine.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts (hunk 8 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts:237]: It is not taking 3 seconds. I added this randomly removed this and update the test case.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.spec.ts (hunk 2 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.spec.ts:150]: Updated

**Review by hmoreras — APPROVED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts [hunk 11]
```diff
@@ -158,5 +209,37 @@ describe('DotContainerPropertiesComponent', () => {
...
+            fixture.componentInstance.form.get('maxContentlets').setValue(2);
+            const contentTypeButton = de.query(By.css('[data-testId=showContentTypeAndCode]'));
+            contentTypeButton.triggerEventHandler('click');
+            tick();
+            const preLoopComponent = de.query(By.css('dot-loop-editor'));
+            const codeEditoromponent = de.query(By.css('dot-container-code'));
+            expect(contentTypeButton.attributes.disable).not.toBeDefined();
+            expect(preLoopComponent).toBeDefined();
+            expect(codeEditoromponent).toBeDefined();
+        }));
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts:243]: You need to test the users actions here.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts (hunk 7 above)
Comment (fmontes) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts]: Check this doc please, Angular have a specific way to testing events.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.ts [hunk 12]
```diff
@@ -59,16 +54,46 @@ export class DotContainerPropertiesComponent implements OnInit {
...
+                            containerInode: new FormControl(containerInode),
+                            contentTypeVar: new FormControl(contentTypeVar)
+                        })
+                    );
+                }
+            );
+
+            // Using setTimeout because dotGlobal Message Service throwing ExpressionChangedAfterItHasBeenCheckedError Error
+            setTimeout(() => {
+                this.showContentTypeAndCode();
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.ts:92]: Can you please check if why we need to call the `dotGlobalMessageService` inside the store?

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts (hunk 7 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts]: I already updated it accordingly. would you please review it?

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.ts (hunk 12 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.ts:92]: I also updated you about this on slack. basically It is showing a message when it is getting content Types from API but I don't think it is necessary.
image

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.ts (hunk 9 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.ts]: updated

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts (hunk 5 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts]: updated

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts (hunk 4 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts]: updated

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts (hunk 3 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.ts]: updated

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts (hunk 7 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts]: Would you please check again it doesn't return `EventEmitter` so we don't need to subscribe event it should simply click and called the function which shows the container content types.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.spec.ts [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+    describe('with data', () => {
+        beforeEach(fakeAsync(() => {
+            spyOn<CoreWebService>(coreWebService, 'requestView').and.returnValue(
+                of({
+                    entity: mockContentTypes
+                })
+            );
+            hostFixture.detectChanges();
+            tick();
+            hostFixture.detectChanges();
```
[the remark this answers is not in the document]
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.spec.ts]: updated

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.spec.ts [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+            menu = de.query(By.css('p-menu')).componentInstance;
+            menu.model[0].command();
+            hostFixture.detectChanges();
+            const contentTypes = de.queryAll(By.css('p-tabpanel'));
+            expect(contentTypes.length).toEqual(3);
+            expect((hostComponent.form.get('containerStructures') as FormArray).length).toEqual(2);
+        });
+
+        it('should have remove content type', () => {
+            menu = de.query(By.css('p-menu')).componentInstance;
```
[the remark this answers is not in the document]
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-code/dot-container-code.component.spec.ts]: updated

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts (hunk 11 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.spec.ts:246]: updated

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.ts (hunk 12 above)
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/dot-container-properties.component.ts]: updated

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/store/dot-container-properties.store.ts [hunk 15]
```diff
@@ -167,17 +167,13 @@ export class DotContainerPropertiesStore extends ComponentStore<DotContainerProp
...
                 return this.dotContentTypeService.getContentTypes({ page: 999 });
             }),
             tap((contentTypes: DotCMSContentType[]) => {
-                this.dotGlobalMessageService.success(this.dotMessageService.get('loaded'));
                 this.updateContentTypes(contentTypes);
                 this.updateContentTypeVisibility(true);
             }),
             catchError((err: HttpErrorResponse) => {
-                this.dotGlobalMessageService.error(err.statusText);
                 this.dotHttpErrorManagerService.handle(err);
```
[the remark this answers is not in the document]
Comment (zulqarnainvd) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-containers/dot-container-create/dot-container-properties/store/dot-container-properties.store.ts:177]: ok, got it. I assumed as per our discussion I only needed to remove it from the ContentType API call.

**Pull request merged**